Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,265 @@
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run(file: String = #file, line: Int = #line) async throws {
+        switch self.state {
+        case .initial:
+            self.state = .running
+            try await self._run()
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
+
+            case .running:
+                self.state = .finished
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning(file: file, line: line)
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished(file: file, line: line)
+        }
+    }
+
+    private func _run() async throws {
+        self.logger.info(
+            "Starting service lifecycle",
+            metadata: [
+                self.configuration.logging.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
+                self.configuration.logging.servicesKey: "\(self.services)",
+            ]
+        )
+
+        enum ChildTaskResult {
+            case serviceFinished(service: any Service, index: Int)
+            case serviceThrew(service: any Service, index: Int, error: any Error)
+            case signalCaught(UnixSignal)
+            case signalSequenceFinished
+        }
+
+        // Using a result here since we want a task group that has non-throwing child tasks
+        // but the body itself is throwing
+        let result = await withTaskGroup(of: ChildTaskResult.self, returning: Result<Void, Error>.self) { group in
+            // First we have to register our signals.
+            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
+
+            // We have to create a graceful shutdown manager per service
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                group.addTask {
```

Comment:
can we add a comment here that this must remain addTask and should _not_ become `addTaskUnlessCancelled` to ensure that we close all services :)